### ISSUE TITLE
`SX` `solve`/`inv` return silent zeros for `n <= 3` on a matrix with condition number 1.3

### ISSUE THREAD

**Issue by shadow-orange41** [casadi/casadi#4382]
Take `M = eye(3) + 0.1`, `b = ones(3)`: `cond(M) = 1.3`, and `solve` is
homogeneous of degree 0, so `solve(M*s, b*s)` is `0.769230769230769...` at
every scale `s`. At `s = 1e103`, `SX` returns

```
  SX    solve[0]  0
  SX    inv[0][0] 0
```

both of which are incorrect: the exact values are `0.769230769230769...` and
`9.23e-104`, ordinary doubles that `SX` reports as exact zeros with no error,
no warning, and no non-finite value to signal it.

`numpy`, `torch`, `jax`, and casadi's *own* `MX` `solve` all return the correct
value on every input below.

## Minimal reproducer (C++)

```cpp
#include <casadi/casadi.hpp>
#include <cstdio>
#include <vector>

using namespace casadi;

int main() {

  const double s = 1e103;
  DM M = (DM::eye(3) + 0.1) * s, b = DM::ones(3, 1) * s;

  SX A = SX::sym("A", 3, 3), v = SX::sym("v", 3, 1);
  Function sx_solve("sx_solve", {A, v}, {solve(A, v)});
  Function sx_inv("sx_inv", {A}, {inv(A)});

  MX Am = MX::sym("A", 3, 3), vm = MX::sym("v", 3, 1);
  Function mx_solve("mx_solve", {Am, vm}, {solve(Am, vm)});

  std::printf("casadi %s   n = 3, cond(M) = 1.3, scale s = %.0e\n\n",
              CasadiMeta::version(), s);
  std::printf("  exact solve[0]  %.17g\n", 0.76923076923076927);
  std::printf("  SX    solve[0]  %.17g\n",
              static_cast<double>(sx_solve(std::vector<DM>{M, b})[0](0)));
  std::printf("  SX    inv[0][0] %.17g\n",
              static_cast<double>(sx_inv(std::vector<DM>{M})[0](0, 0)));
  std::printf("  MX    solve[0]  %.17g\n",
              static_cast<double>(mx_solve(std::vector<DM>{M, b})[0](0)));
  return 0;
}
```

### Output

```
casadi 3.7.2+   n = 3, cond(M) = 1.3, scale s = 1e+103

  exact solve[0]  0.76923076923076927
  SX    solve[0]  0
  SX    inv[0][0] 0
  MX    solve[0]  0.76923076923076938
```

## Same inputs in torch and jax

```python
import torch

s = 1e103
M = (torch.eye(3, dtype=torch.float64) + 0.1) * s
b = torch.ones(3, 1, dtype=torch.float64) * s

print(f"torch {torch.__version__}   n = 3, cond(M) = 1.3, scale s = {s:.0e}\n")
print(f"  exact solve[0]  {0.76923076923076927!r}")
print(f"  torch solve[0]  {torch.linalg.solve(M, b)[0, 0].item()!r}")
print(f"  torch inv[0][0] {torch.linalg.inv(M)[0, 0].item()!r}")
```

### Output

```
torch 2.14.0.dev20260809+cpu   n = 3, cond(M) = 1.3, scale s = 1e+103

  exact solve[0]  0.7692307692307693
  torch solve[0]  0.769230769230769
  torch inv[0][0] 9.230769230769228e-104

jax 0.11.1.dev20260810   n = 3, cond(M) = 1.3, scale s = 1e+103

  exact solve[0]  0.7692307692307693
  jax   solve[0]  0.7692307692307692
  jax   inv[0][0] 9.230769230769228e-104
```

`9.23e-104` being small is not a reason to accept `0` for it: the smallest
normal double is `2.2e-308`, so the entry carries a full 53-bit mantissa, and
its magnitude is set by the scale of the input, not by any loss of information.

`n >= 4` is correct where `n = 3` is not.

**Comment by apozharski** [comment]
Root cause here btw @jgillis is because for `n<=3` the explicit determinant of `A` is formed which is not representable by a double, as it is $1.3 \times 10^{309}$ and is "rounded" to `+inf`. The reason that the `MX::solve` works is I believe it falls back to the QR factorization. The reporter seems to think that condition number is the only thing that determines the accuracy of linear algebra routines, which is false, poorly scaled problems (whether due to scale mismatch, or due to representation limits) will behave poorly unless the user is careful about exactly what algorithm is used. 

https://github.com/casadi/casadi/blob/973b086f4dcda9f49cd9c1948432ae4b7ee54886/casadi/core/matrix_impl.hpp#L2445-L2450

**Comment by shadow-orange41** [comment]
Thanks for tracking down the root cause. That makes sense, and I take the point that condition number alone doesn't guarantee a given algorithm stays well-scaled internally.

That said, I think the overflow itself is the more actionable problem here, separate from the conditioning discussion. `inv_minor` computes a determinant that overflows to `+inf`, and that `+inf` then silently turns into a `0`.

Would you want to put up a PR for this?@apozharski If not, I'm happy to — but I'd like the developers' take on the direction first. A few options come to my mind:

- Making the determinant/minor-expansion computation more numerically stable
- Checking for `isinf`/`isnan` on the determinant (or final result) and raising an error when it fires
- Falling back to QR for large-magnitude inputs, similar to what `MX::solve` already does

Let me know which direction seems best, or if there's another approach you'd prefer.

**Comment by apozharski** [comment]
> Making the determinant/minor-expansion computation more numerically stable

This is nonsense. This is not a stability issue, the _correct_ determinant is just not representable by a IEEE754 double precision float.

> `inv_minor` computes a determinant that overflows to +inf, and that +inf then silently turns into a 0.

This is also correct behavior. The IEEE754 standard states that division of a finite dividend by an infinite divisor yield zero of the sign corresponding to the sign of the divisor.

> Falling back to QR for large-magnitude inputs, similar to what MX::solve already does

`MX::solve` always defaults to QR for generic matrices, as far as I can tell (but maybe this is actually the best option here). This also would require you to choose what "large" means  or would basically require calculating the determinant anyway.

As for the last remaining option: I personally think adding an additional branch (which albeit on most x86 machines, in many cases, will be correctly predicted, though not necessarily so on embedded hardware) and forcing every user to pay its cost, seems silly. Especially this is because such a small matrix solve could actually find itself being run in a hot loop. The user can themselves call `qr_solve` if they know that the determinant is likely to be large, and the rows sufficiently far from being linearly dependent. Then again I don't speak for Joris and Joel, so the final decision is up to them, and this is just my 0.02€.
